File: JavaScript/src/GildedRose.js
import AgingStrategies from './AgingStrategies';
import Item from './Item';

// Handles inventory updates over time
export default class GildedRose {
  constructor({ name, daysRemaining, quality }) {
    this.item = new Item({ name, daysRemaining, quality });
  }

  get name() {
    return this.item.name;
  }

  get daysRemaining() {
    return this.item.daysRemaining;
  }

  get quality() {
    return this.item.quality;
  }

  tick() {
    const strategyKlass = AgingStrategies.strategyFor(this.item.name);
    this.item.tick(strategyKlass);
  }
}

File: Ruby/lib/gilded_rose.rb
# frozen_string_literal: true

require 'item'
require 'decorators/aged'
require 'decorators/normal'
require 'decorators/null'
require 'decorators/time_sensitive'

# Factory for items decorated with aging behavior
class GildedRose
  def self.build(name:, days_remaining:, quality:)
    item = Item.new(
      name: name,
      days_remaining: days_remaining,
      quality: quality
    )
    decorator_klass = decorator_for(name)
    decorator_klass.new(item)
  end

  private_class_method def self.decorator_for(item_name)
    {
      'Aged Brie' => Decorators::Aged,
      'Sulfuras, Hand of Ragnaros' => Decorators::Null,
      'Backstage passes to a TAFKAL80ETC concert' => Decorators::TimeSensitive
    }[item_name] || Decorators::Normal
  end
end
